### PR TITLE
Issue #12011: tweet-releasenotes.sh should be able to work on start and finish release names.

### LINKED ISSUE #12011: tweet-releasenotes.sh should be able to generate release notes for any release of past
https://github.com/checkstyle/checkstyle/blob/master/.ci/tweet-releasenotes.sh

1)
it is not clear why we demand 
`checkForVariable "CS_RELEASE_VERSION"`

and in same time do
```
CS_RELEASE_VERSION="$(getCheckstylePomVersion)"
echo CS_RELEASE_VERSION="$CS_RELEASE_VERSION"
```

2)
This script should work in two modes:
`.ci/tweet-releasenotes.sh` - generation of twit for latest release
`.ci/tweet-releasenotes.sh 10.3.2` - generation of twit for release 10.3.2

### PR DESCRIPTION
**Pull request by Rahulkhinchi03**
Fixes #12011

### CODE REVIEW

**Comment by romani** [comment]
migrate all release notes scripts to work for any release of past - https://github.com/checkstyle/checkstyle/pull/12007#issuecomment-1217268954

File: .ci/tweet-releasenotes.sh [hunk 1]
```diff
@@ -32,17 +32,28 @@ CS_RELEASE_VERSION="$(getCheckstylePomVersion)"
 echo CS_RELEASE_VERSION="$CS_RELEASE_VERSION"
 
 cd .ci-temp/checkstyle
-LATEST_RELEASE_TAG=$(curl -s https://api.github.com/repos/checkstyle/checkstyle/releases/latest \
-                       | jq ".tag_name")
-echo LATEST_RELEASE_TAG="$LATEST_RELEASE_TAG"
+
+curl https://api.github.com/repos/checkstyle/checkstyle/releases -o /var/tmp/cs-releases.json
```
Comment (romani) [inline .ci/tweet-releasenotes.sh:36]: add auth token

File: .ci/tweet-releasenotes.sh [hunk 2]
```diff
@@ -32,17 +32,28 @@ CS_RELEASE_VERSION="$(getCheckstylePomVersion)"
...
+TARGET_RELEASE_INDEX=$(cat /var/tmp/cs-releases.json | \
+    jq '[.[].tag_name] | to_entries | .[] | select(.value=="checkstyle-$1") | .key')
```
Comment (romani) [inline .ci/tweet-releasenotes.sh:38]: better to give better name to `$1`, use extra variable

**Review by romani — CHANGES_REQUESTED** [review]
share test results

items:

File: .ci/tweet-releasenotes.sh (hunk 2 above)
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh:38]: Changes Done

File: .ci/tweet-releasenotes.sh (hunk 1 above)
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh:36]: Changes Done

File: .ci/tweet-releasenotes.sh [hunk 3]
```diff
@@ -32,17 +32,32 @@ CS_RELEASE_VERSION="$(getCheckstylePomVersion)"
...
+
+LATEST_RELEASE=$1
+TARGET_RELEASE_INDEX=$(cat /var/tmp/cs-releases.json | \
+    jq '[.[].tag_name] | to_entries | .[] | select(.value=="checkstyle-$LATEST_RELEASE") | .key')
+echo "$TARGET_RELEASE_INDEX"
+PREVIOUS_RELEASE_INDEX=$(($TARGET_RELEASE_INDEX+1))
+echo "$PREVIOUS_RELEASE_INDEX"
+
+START_REF=$(cat /var/tmp/cs-releases.json | jq -r ".[$TARGET_RELEASE_INDEX].tag_name")
+END_REF=$(cat /var/tmp/cs-releases.json | jq -r ".[$PREVIOUS_RELEASE_INDEX].tag_name")
```
Comment (romani) [inline .ci/tweet-releasenotes.sh:48]: END_REF should be taken by TARGET_RELEASE_INDEX.

builder goes from start(old release) to end (target release).

**Review by romani — CHANGES_REQUESTED** [review]
items

File: .ci/tweet-releasenotes.sh (hunk 1 above)
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh]: I am not sure if adding an `AUTH TOKEN` is a good idea. 

My API request never got rejected without `AUTH TOKEN` but with `AUTH TOKEN`, sometimes my expected outcome is not the same.

Example:
Screenshot from 2022-08-17 19-19-39


Code, if want to test:

```
 curl \
 https://api.github.com/repos/checkstyle/checkstyle/releases \
 -o /var/tmp/cs-releases.json

TARGET_RELEASE_NUM=10.1
TARGET_RELEASE_INDEX=$(cat /var/tmp/cs-releases.json | \
    jq '[.[].tag_name] | to_entries | .[] | select(.value=="checkstyle-'$TARGET_RELEASE_NUM'") | .key')
echo "$TARGET_RELEASE_INDEX"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  115k  100  115k    0     0   364k      0 --:--:-- --:--:-- --:--:--  367k
```
OUTPUT: `4`

```
 curl \
 https://api.github.com/repos/checkstyle/checkstyle/releases \
 -H "Authorization: token ghp_..." \
 -o /var/tmp/cs-releases.json

TARGET_RELEASE_NUM=10.1
TARGET_RELEASE_INDEX=$(cat /var/tmp/cs-releases.json | \
    jq '[.[].tag_name] | to_entries | .[] | select(.value=="checkstyle-'$TARGET_RELEASE_NUM'") | .key')
echo "$TARGET_RELEASE_INDEX"
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    90  100    90    0     0    158      0 --:--:-- --:--:-- --:--:--   158
```

ERROR: `jq: error (at <stdin>:4): Cannot index string with string "tag_name"`

File: .ci/tweet-releasenotes.sh (hunk 3 above)
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh:48]: Changes Done

File: .ci/tweet-releasenotes.sh [hunk 4]
```diff
@@ -32,17 +32,32 @@ CS_RELEASE_VERSION="$(getCheckstylePomVersion)"
...
 cd .ci-temp/checkstyle
-LATEST_RELEASE_TAG=$(curl -s https://api.github.com/repos/checkstyle/checkstyle/releases/latest \
-                       | jq ".tag_name")
-echo LATEST_RELEASE_TAG="$LATEST_RELEASE_TAG"
+
+curl https://api.github.com/repos/checkstyle/checkstyle/releases \
+    -H "Authorization: token $GITHUB_TOKEN" \
+    -o /var/tmp/cs-releases.json
+
+LATEST_RELEASE=$1
```
[the remark this answers is not in the document]
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh:40]: Changes Done

**Comment by Rahulkhinchi03** [comment]
TESTING:

```
rahul@rk7:~/Desktop/opensource/checkstyle$ ./.ci/tweet-releasenotes.sh 10.1
Cloning into 'contribution'...
remote: Enumerating objects: 7947, done.
...
...
...
Cloning into 'checkstyle'...
remote: Enumerating objects: 210627, done.
remote: Counting objects: 100% (420/420), done.
remote: Compressing objects: 100% (222/222), done.
remote: Total 210627 (delta 126), reused 322 (delta 87), pack-reused 210207
Receiving objects: 100% (210627/210627), 82.33 MiB | 3.48 MiB/s, done.
Resolving deltas: 100% (120563/120563), done.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  115k    0  115k    0     0   119k      0 --:--:-- --:--:-- --:--:--  119k
4
5
START_REF=checkstyle-10.0
END_REF=checkstyle-10.1

[WARN] Issue #6207 "Expand XPath IT Regression Testing" is not closed! Please review issue https://github.com/checkstyle/checkstyle/issues/6207
[WARN] Issue #10924 "All checks that use `AbstractCheck#getLine()` should check code points instead of characters for spacing" is not closed! Please review issue https://github.com/checkstyle/checkstyle/issues/10924
[WARN] Issue #11446 "Update Tests to use new 'verifyXxxxxx' method or 'execute' that use inlined config in Input files" is not closed! Please review issue https://github.com/checkstyle/checkstyle/issues/11446
[WARN] Issue #11214 "Specify violation messages in input files." is not closed! Please review issue https://github.com/checkstyle/checkstyle/issues/11214
[WARN] Issue #4845 "Test inputs should be completely standalone." is not closed! Please review issue https://github.com/checkstyle/checkstyle/issues/4845
[WARN] Issue #11166 "Fix existing cases of `getFileContents()` usage" is not closed! Please review issue https://github.com/checkstyle/checkstyle/issues/11166

Execution succeeded!
rahul@rk7:~/Desktop/opensource/checkstyle$ 
```

GENERATION:
Screenshot from 2022-08-17 19-40-10

File: .ci/tweet-releasenotes.sh (hunk 1 above)
Comment (romani) [reply .ci/tweet-releasenotes.sh]: >Cannot index string with string "tag_name"

it means that response was not a json, please execute command few times without `| jq` to see that response you got and please share it here.

File: .ci/tweet-releasenotes.sh [hunk 5]
```diff
@@ -28,22 +28,35 @@ else
...
+
+curl \
+ https://api.github.com/repos/checkstyle/checkstyle/releases \
+ -o /var/tmp/cs-releases.json
+
+TARGET_RELEASE_NUM=$1
+TARGET_RELEASE_INDEX=$(cat /var/tmp/cs-releases.json | \
+    jq '[.[].tag_name] | to_entries | .[] | \
+    select(.value=="checkstyle-'"$TARGET_RELEASE_NUM"'") | .key')
+echo "$TARGET_RELEASE_INDEX"
```
Comment (romani) [inline .ci/tweet-releasenotes.sh:41]: print variable name to make it read log output clearly without reading a code.
do it in both places.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .ci/tweet-releasenotes.sh [hunk 6]
```diff
@@ -28,22 +28,35 @@ else
...
-echo LATEST_RELEASE_TAG="$LATEST_RELEASE_TAG"
+
+curl \
+ https://api.github.com/repos/checkstyle/checkstyle/releases \
+ -o /var/tmp/cs-releases.json
+
+TARGET_RELEASE_NUM=$1
+TARGET_RELEASE_INDEX=$(cat /var/tmp/cs-releases.json | \
+    jq '[.[].tag_name] | to_entries | .[] | \
+    select(.value=="checkstyle-'"$TARGET_RELEASE_NUM"'") | .key')
```
Comment (romani) [inline .ci/tweet-releasenotes.sh:40]: please make it like:
```
TARGET_RELEASE_INDEX=$(cat /var/tmp/cs-releases.json | \
    jq "[.[].tag_name] | to_entries | .[] | \
    select(.value==\"checkstyle-$TARGET_RELEASE_NUM\") | .key")
```

I feel uncertain  with jq and combined `''` strings ... it works, but it not sure why, lets avoid this style for now.

**Review by romani — CHANGES_REQUESTED** [review]
items:

File: .ci/tweet-releasenotes.sh (hunk 5 above)
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh:41]: Changes Done

File: .ci/tweet-releasenotes.sh (hunk 6 above)
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh:40]: Changes Done

File: .ci/tweet-releasenotes.sh (hunk 1 above)
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh]: Works fine now. The error was coming from the quotes at `select(.value=="checkstyle-'$TARGET_RELEASE_NUM'"`

**Review by romani — APPROVED** [review]
Ok to merge

File: .ci/tweet-releasenotes.sh [hunk 7]
```diff
@@ -28,22 +28,37 @@ else
...
-echo CS_RELEASE_VERSION="$CS_RELEASE_VERSION"
-
 cd .ci-temp/checkstyle
-LATEST_RELEASE_TAG=$(curl -s https://api.github.com/repos/checkstyle/checkstyle/releases/latest \
-                       | jq ".tag_name")
-echo LATEST_RELEASE_TAG="$LATEST_RELEASE_TAG"
+
+curl \
+ https://api.github.com/repos/checkstyle/checkstyle/releases \
+ -H "Authorization: token $GITHUB_TOKEN" \
```
Comment (romani) [inline .ci/tweet-releasenotes.sh:35]: GITHUB_READ_ONLY_TOKEN

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .ci/tweet-releasenotes.sh (hunk 7 above)
Comment (Rahulkhinchi03) [reply .ci/tweet-releasenotes.sh]: Changes Done

**Review by romani — APPROVED** [review]
Ok to merge

**Pull request merged**